### ISSUE TITLE
[REQ] External DB Support

### ISSUE THREAD

**Issue by vitospaghetti** [giuseppe99barchetta/SuggestArr#93]
**Is your feature request related to a problem? Please describe.**
Looking for easy support within Kubernetes. Config file should be easy to create dynamically, but the requests.db would be nice to have in external DB like Postgres (preferred) or MariaDB

**Describe the solution you'd like**
Generic DB layer to support SQLite, Postgres or MariaDB

**Describe alternatives you've considered**
N/A

**Comment by giuseppe99barchetta** [comment]
Hi! Could you help me test this feature? It's currently deployed under the **nightly** tag in Docker.

**Comment by gottsman** [comment]
I did a quick default install with Postgres and the setup seemed to work as expected. I did get the wizard page asking for DB information even though I populated the DB_HOST,DB_PORT..... environment variables in a quick docker-compose file. I'll also try MariaDB as well.

Suggestarr is hands-off so I don't generally see errors. Not sure if this is a related bug or not. Error appeared in a browser pop-up and the logs.

`api_service.blueprints.automation.routes - ERROR - Unexpected error: 'ContentAutomation' object has no attribute 'media_handler'`

Definitely a huge step forward....

**Comment by giuseppe99barchetta** [comment]
> I did a quick default install with Postgres and the setup seemed to work as expected. I did get the wizard page asking for DB information even though I populated the DB_HOST,DB_PORT..... environment variables in a quick docker-compose file. I'll also try MariaDB as well.
> 
> Suggestarr is hands-off so I don't generally see errors. Not sure if this is a related bug or not. Error appeared in a browser pop-up and the logs.
> 
> `api_service.blueprints.automation.routes - ERROR - Unexpected error: 'ContentAutomation' object has no attribute 'media_handler'`
> 
> Definitely a huge step forward....

Hi! Thank you for the quick response!

Just to clarify, Suggestarr doesn’t currently support environment variable configuration for this. When you fill in the fields from the interface, it generates a config.yaml file inside the volume specified in the Docker Compose setup. This file stores all the necessary configuration.

Could you please let me know under which circumstances the error appeared? Also, if you manually trigger the content request, do you see the tables being created and populated correctly?

Thanks again!

**Comment by giuseppe99barchetta** [comment]
> Cleaned out the environment variables for now. I'll look to see if there is a feature request for automating the wizard with variables.
> 
> Ran the wizard a few times now and am not getting any data in the DB. I don't see any tables being created.
> 
> First as suggestarr user with "Create" Role, then "Super User" role and then logged in with Postgres root DB user. No luck. Manually tried "Run now" even with PG admin user and nothing was created, but it did complete.
> 
> Checking the ERROR logs revealed this...
> 
> `2025-02-07 16:27:00,243 - api_service.blueprints.automation.routes - ERROR - Unexpected error: DatabaseManager.handle_sql_error() takes 3 positional arguments but 4 were given`
> 
> Here is the relevant config.yaml portion:
> 
> ```
> DB_HOST: 192.168.3.2
> DB_NAME: suggestarr
> DB_PASSWORD: <redacted>
> DB_PORT: '5432'
> DB_TYPE: postgres
> DB_USER: postgres
> ```
> 
> Is there a way to increase the DEBUG logs?

I need to run additional tests with Postgres. I'll get back to you with a fix later, or at the latest by tomorrow.

**Comment by giuseppe99barchetta** [comment]
> Cleaned out the environment variables for now. I'll look to see if there is a feature request for automating the wizard with variables.
> 
> Ran the wizard a few times now and am not getting any data in the DB. I don't see any tables being created.
> 
> First as suggestarr user with "Create" Role, then "Super User" role and then logged in with Postgres root DB user. No luck. Manually tried "Run now" even with PG admin user and nothing was created, but it did complete.
> 
> Checking the ERROR logs revealed this...
> 
> `2025-02-07 16:27:00,243 - api_service.blueprints.automation.routes - ERROR - Unexpected error: DatabaseManager.handle_sql_error() takes 3 positional arguments but 4 were given`
> 
> Here is the relevant config.yaml portion:
> 
> ```
> DB_HOST: 192.168.3.2
> DB_NAME: suggestarr
> DB_PASSWORD: <redacted>
> DB_PORT: '5432'
> DB_TYPE: postgres
> DB_USER: postgres
> ```
> 
> Is there a way to increase the DEBUG logs?

Hi! I've pushed an update with some fixes for PostgreSQL. Would you have time to test it? It's available under the nightly tag.

**Comment by gottsman** [comment]
> Hi! I've pushed an update with some fixes for PostgreSQL. Would you have time to test it? It's available under the nightly tag.

Pulled the latest image from about 4 hours ago. I am still not seeing tables being created in the Postgres DB. I tried as the suggestarr user as well as the PG root user. 

Another issue I found was requests being generated as my admin user (and subsequently auto accepted) instead of the suggestarr user. I entered the suggestarr user details and verified the connection during the setup wizard. I noticed this from the previous nightly build as well.

**Comment by gottsman** [comment]
> It's strange, on my local PostgreSQL configuration, the table was created correctly, and no issues were detected. Give me some time to figure out what might be causing this. Did you encounter any errors when forcing the run? If so, could you share the log here?

Here is the relevant log... I don't see much related to DB activity.

```
2025-02-10 13:57:37,656 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2025-02-10 13:57:37,658 INFO supervisord started with pid 1
2025-02-10 13:57:38,661 INFO spawned: 'uvicorn' with pid 7
2025-02-10 13:57:39,610 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 13:57:39,637 - api_service.utils.utils - INFO - 
        
        =====================================================================================
        |   Welcome to the SuggestArr Application!                                          |
        |   Manage your settings through the web interface at: http://localhost:5000        |
        |   Fill in the input fields with your data and let the cron job handle the rest!   |
        |   To run the automation process immediately, click the 'Force Run' button.        |
        |   The 'Force Run' button will appear only after you save your settings.           |
        |   To leave feedback visit: https://github.com/giuseppe99barchetta/SuggestArr      |
        =====================================================================================
        
2025-02-10 13:57:39,651 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 13:57:39,656 - api_service.config.cron_jobs - INFO - Cron job set with expression: 0 0 * * *
2025-02-10 13:57:40,657 INFO success: uvicorn entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2025-02-10 13:59:08,970 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 13:59:08,980 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 13:59:26,344 - JellyfinClient - INFO - Searching Jellyfin libraries.
2025-02-10 13:59:54,346 - SeerClient - ERROR - Login failed: 403
2025-02-10 14:00:31,246 - api_service.config.config - INFO - Creating new file for config at /app/config/config_files/config.yaml
2025-02-10 14:00:31,254 - api_service.config.cron_jobs - INFO - Cron job set with expression: 0 0 * * *
2025-02-10 14:01:09,535 - SeerClient - INFO - Fetched all requests and saved to database.
2025-02-10 14:01:09,536 - JellyfinClient - INFO - Searching all content in Jellyfin
2025-02-10 14:03:11,696 - JellyfinClient - INFO - Retrieved 6961 items in Movies
2025-02-10 14:03:20,270 - JellyfinClient - INFO - Retrieved 497 items in Shows
2025-02-10 14:03:20,274 - ContentAutomation - INFO - Fetching content for user: <redacted>
2025-02-10 14:03:20,275 - ContentAutomation - INFO - Fetching content for user: <redacted>
2025-02-10 14:03:20,276 - ContentAutomation - INFO - Fetching content for user: <redacted>
2025-02-10 14:03:20,276 - ContentAutomation - INFO - Fetching content for user: <redacted>
2025-02-10 14:03:20,663 - JellyfinClient - INFO - Retrieved 2 watched items in Movies. for user <redacted>
2025-02-10 14:03:20,664 - JellyfinClient - INFO - Retrieved 3 watched items in Movies. for user <redacted>
2025-02-10 14:03:20,762 - JellyfinClient - INFO - Retrieved 10 watched items in Movies. for user <redacted>
2025-02-10 14:03:20,763 - JellyfinClient - INFO - Retrieved 10 watched items in Movies. for user <redacted>
2025-02-10 14:03:22,471 - JellyfinClient - INFO - Retrieved 2 watched items in Shows. for user <redacted>
2025-02-10 14:03:22,472 - JellyfinClient - INFO - Retrieved 2 watched items in Shows. for user <redacted>
2025-02-10 14:03:22,473 - JellyfinClient - INFO - Retrieved 2 watched items in Shows. for user <redacted>
2025-02-10 14:03:22,474 - JellyfinClient - INFO - Retrieved 3 watched items in Shows. for user <redacted>
...... More INFO messages from TMDBClient and ContentAutomation
```

**Comment by giuseppe99barchetta** [comment]
> 2025-02-10 13:59:54,346 - SeerClient - ERROR - Login failed: 403

As i can see in the logs probably there's a misconfiguration for the Seer local account. I've pushed another update, now you can enable in environment variables **LOG_LEVEL=debug** to print db related logging (actually only on create tables) so we can understand what happening in your docker.

Please update the container and paste here the log

**Comment by gottsman** [comment]
Interesting logs....

1. For the DB portion, I tried with the suggestarr user and PG root user. Both authenticated successfully and the DB is already created. Its weird that it still tries to use SQLite.
2. I was able to authenticate to Jellyyseer with the API key. I was able to pick the suggestarr user from the dropdown and enter the *incorrect* password. No on-screen errors were reported. It let me proceed to the DB portion of the wizard without proper credentials for the suggestarr user.
3. I subsequently changed the password and entered it properly and it said "Logged In" as expected. I think it would be better to generate a visual error on an errant password.


```
2025-02-10 16:02:15,025 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2025-02-10 16:02:15,028 INFO supervisord started with pid 1
2025-02-10 16:02:16,032 INFO spawned: 'uvicorn' with pid 7
2025-02-10 16:02:16,920 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 16:02:16,946 - api_service.app - INFO - Current log level: DEBUG
2025-02-10 16:02:16,947 - api_service.utils.utils - INFO -

        =====================================================================================
        |   Welcome to the SuggestArr Application!                                          |
        |   Manage your settings through the web interface at: http://localhost:5000        |
        |   Fill in the input fields with your data and let the cron job handle the rest!   |
        |   To run the automation process immediately, click the 'Force Run' button.        |
        |   The 'Force Run' button will appear only after you save your settings.           |
        |   To leave feedback visit: https://github.com/giuseppe99barchetta/SuggestArr      |
        =====================================================================================

2025-02-10 16:02:16,962 - api_service.utils.utils - DEBUG - Environment variables reloaded.
2025-02-10 16:02:16,962 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
2025-02-10 16:02:16,970 - api_service.config.cron_jobs - INFO - Cron job set with expression: 0 0 * * *
INFO:     Started server process [7]
INFO:     Waiting for application startup.
INFO:     ASGI 'lifespan' protocol appears unsupported.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:5000 (Press CTRL+C to quit)
INFO:     172.16.3.3:33500 - "GET / HTTP/1.1" 304 Not Modified
2025-02-10 16:02:17,307 INFO success: uvicorn entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
INFO:     172.16.3.3:33500 - "GET /js/chunk-vendors.627d225a.js HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33512 - "GET /js/app.b6a0dbd8.js HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33528 - "GET /css/app.fe38e01a.css HTTP/1.1" 304 Not Modified
2025-02-10 16:02:17,360 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
INFO:     172.16.3.3:33528 - "GET /api/config/fetch HTTP/1.1" 200 OK
INFO:     172.16.3.3:33528 - "GET /img/logo.8e3a4916.png HTTP/1.1" 304 Not Modified
2025-02-10 16:02:17,374 - api_service.config.config - WARNING - /app/config/config_files/config.yaml not found. Creating a new one with default values.
INFO:     172.16.3.3:33512 - "GET /api/config/fetch HTTP/1.1" 200 OK
INFO:     172.16.3.3:33512 - "GET /images/default1.jpg HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33528 - "GET /img/jellyfin-logo.29a3eb64.png HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33500 - "GET /img/plex-logo.67c10e65.png HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33500 - "GET /favicon.ico HTTP/1.1" 304 Not Modified
INFO:     172.16.3.3:33500 - "GET / HTTP/1.1" 200 OK
INFO:     172.16.3.3:33500 - "GET / HTTP/1.1" 200 OK
2025-02-10 16:02:34,646 - JellyfinClient - INFO - Searching Jellyfin libraries.
INFO:     172.16.3.3:50052 - "POST /api/jellyfin/libraries HTTP/1.1" 200 OK
INFO:     172.16.3.3:50052 - "POST /api/jellyfin/users HTTP/1.1" 200 OK
INFO:     172.16.3.3:51878 - "POST /api/seer/get_users HTTP/1.1" 200 OK
2025-02-10 16:03:10,793 - SeerClient - INFO - Successfully logged in as suggestarr
INFO:     172.16.3.3:48782 - "POST /api/seer/login HTTP/1.1" 200 OK
2025-02-10 16:03:15,651 - SeerClient - ERROR - Login failed: 403
INFO:     172.16.3.3:48782 - "POST /api/seer/login HTTP/1.1" 401 Unauthorized
INFO:     172.16.3.3:48782 - "GET / HTTP/1.1" 200 OK
2025-02-10 16:03:19,210 - SeerClient - INFO - Successfully logged in as suggestarr
INFO:     172.16.3.3:48782 - "POST /api/seer/login HTTP/1.1" 200 OK
2025-02-10 16:03:51,688 - DatabaseManager - DEBUG - Using SQLite database.
2025-02-10 16:03:51,698 - DatabaseManager - DEBUG - Requests table already exists.
2025-02-10 16:03:51,702 - DatabaseManager - DEBUG - Metadata table already exists.
INFO:     172.16.3.3:60364 - "POST /api/config/test-db-connection HTTP/1.1" 200 OK
2025-02-10 16:03:55,380 - api_service.config.config - INFO - Creating new file for config at /app/config/config_files/config.yaml
2025-02-10 16:03:55,395 - api_service.config.cron_jobs - INFO - Cron job set with expression: 0 0 * * *
INFO:     172.16.3.3:60364 - "POST /api/config/save HTTP/1.1" 200 OK
INFO:     172.16.3.3:48290 - "GET / HTTP/1.1" 200 OK
```

**Comment by giuseppe99barchetta** [comment]
> 2025-02-10 16:03:55,380 - api_service.config.config - INFO - Creating new file for config at /app/config/config_files/config.yaml

It appears that there may be an issue with the config file. Have you set up the Compose file correctly?

```
services:
  suggestarr:
    image: ciuse99/suggestarr:latest
    container_name: SuggestArr
    restart: always
    ports:
      - "5000:5000"
    volumes:
      **- <YOUR CONFIG FOLDER>:/app/config/config_files**
    environment:
      - LOG_LEVEL=${LOG_LEVEL:-info}
```

If you navigate to the corresponding config_files folder, do you see the config.yaml file? Is it properly filled out?

**Comment by gottsman** [comment]
> If you navigate to the corresponding config_files folder, do you see the config.yaml file? Is it properly filled out?

Here are the log and config files.

[suggestarr-nightly-compose-with-logs.zip](https://github.com/user-attachments/files/18737363/suggestarr-nightly-compose-with-logs.zip)

**Comment by giuseppe99barchetta** [comment]
> > If you navigate to the corresponding config_files folder, do you see the config.yaml file? Is it properly filled out?
> 
> Here are the log and config files.
> 
> [suggestarr-nightly-compose-with-logs.zip](https://github.com/user-attachments/files/18737363/suggestarr-nightly-compose-with-logs.zip)

I've just created a Discord server for SuggestArr if you'd like to join. I’m unable to replicate your issue, and this thread is starting to get quite long. You can reach me there for further assistance so we can finally fix this issue:

https://discord.gg/cpjBJ5sK

**Comment by gottsman** [comment]
I tried against **chrisbenincasa/tunarr:0.18.11-nvidia** and it seemed to work fine.
